pkg/volumes/service_authentication.go

```diff
@@ -0,0 +1,122 @@
+package volumes
+
+import (
+	"fmt"
+	"os"
+	"strings"
+	"time"
+
+	"github.com/Yelp/paasta-tools-go/pkg/configstore"
+	"gopkg.in/yaml.v2"
+	corev1 "k8s.io/api/core/v1"
+)
+
+const (
+	authenticatingServicesConfigPath    = "/nail/etc/services/authenticating.yaml"
+	jwtServiceAuthConfigName            = "jwt_service_auth"
+	authenticatingServicesCacheDuration = 10 * time.Minute
+	defaultTokenExpirationSeconds       = 3600
+)
+
+type authenticatingServicesConfig struct {
+	Services []string `yaml:"services"`
+}
+
+type jwtServiceAuthTokenSettings struct {
+	Audience          string `json:"audience"`
+	ContainerPath     string `json:"container_path"`
+	ExpirationSeconds int64  `json:"expiration_seconds"`
+}
+
+type jwtServiceAuthConfig struct {
+	TokenSettings jwtServiceAuthTokenSettings `json:"service_auth_token_settings"`
+}
+
+var authenticatingServicesCache map[string]bool
+var authenticatingServicesLastLoaded time.Time
+
+func getAuthenticatingServices(configPath string) (map[string]bool, error) {
+	now := time.Now()
+	if !now.After(authenticatingServicesLastLoaded.Add(authenticatingServicesCacheDuration)) {
+		return authenticatingServicesCache, nil
+	}
+	data, err := os.ReadFile(configPath)
+	if err != nil {
+		return nil, err
+	}
+	var authenticatingServices authenticatingServicesConfig
+	if err = yaml.Unmarshal(data, &authenticatingServices); err != nil {
+		return nil, err
+	}
+	result := make(map[string]bool)
+	for _, service := range authenticatingServices.Services {
+		result[service] = true
+	}
+	authenticatingServicesCache = result
+	authenticatingServicesLastLoaded = now
+	return result, nil
+}
+
+func ServiceRequiresAuthenticationToken(serviceName string) (bool, error) {
+	authenticatingServices, err := getAuthenticatingServices(authenticatingServicesConfigPath)
+	if err != nil {
+		return false, err
+	}
+	_, found := authenticatingServices[serviceName]
+	return found, nil
+}
+
+func formatServiceAccountVolumeName(audience string) string {
+	formatted := strings.Replace(audience, ".", "dot-", -1)
+	formatted = formatMountName(formatted)
+	formatted = fmt.Sprintf("projected-sa--%s", formatted)
+	if len(formatted) > 63 {
+		return formatted[0:63]
```

Comment:
just to confirm: we'll only ever have a single projected SA volume right?
(just making sure this is safe and we don't need to do the usual hash-truncation method we do elsewhere)

Reply:
For this use case, yes, there'll be another attached for the default service account, and one more if the service has an IAM role associated, but clashing with those is effectively impossible, naming conventions are very different.

Reply:
perfect!